src/server.lisp

```diff
@@ -0,0 +1,184 @@
+  ;;
+  ;;   (1) identity frame, null frame, data frame
+  ;;   (2) identity frame, data frame
+  ;;
+  ;; whichever we get, we also have to reply in the same way, so we track the
+  ;; format in addition to the data.
+  (flet ((unpack-foreign-msg-to-bytes (msg)
+           (cffi:foreign-array-to-lisp (pzmq:msg-data msg)
+                                       `(:array :unsigned-char ,(pzmq:msg-size msg)))))
+    (let (identity)
+      (pzmq:with-message msg
+        (pzmq:msg-recv msg receiver)
+        (setf identity (unpack-foreign-msg-to-bytes msg)))
+      (assert (pzmq:getsockopt receiver :rcvmore))
+      (pzmq:with-message msg
+        (pzmq:msg-recv msg receiver)
+        (cond
+          ((pzmq:getsockopt receiver :rcvmore)
+           (assert (zerop (pzmq:msg-size msg)))
+           (pzmq:with-message msg
+             (assert (not (pzmq:getsockopt receiver :rcvmore)))
+             (values
+              identity
+              t
+              (unpack-foreign-msg-to-bytes msg))))
+          (t
+           (values
+            identity
+            nil
+            (unpack-foreign-msg-to-bytes msg))))))))
+
+(defun %push-raw-request (socket identity null-frame? payload)
+  "Pushes a ZMQ reply onto SOCKET.  Takes the following values:
+
+* IDENTITY:    array of (INTEGER 256) that describes the inteded recipient of the reply.  Copy this from the matching %PULL-RAW-REQUEST.
+* NULL-FRAME?: boolean indicating whether the recipient expects an additional null frame after the identity frame. Copy this from the matching %PULL-RAW-REQUEST.
+* PAYLOAD:     array of (INTEGER 256) that houses the raw reply."
+  ;; transmit messages
+  (cffi:with-foreign-objects ((foreign-identity ':unsigned-char (length identity))
+                              (foreign-payload ':unsigned-char (length payload)))
+    (dotimes (j (length identity))
+      (setf (cffi:mem-aref foreign-identity ':unsigned-char j)
+            (aref identity j)))
+    (dotimes (j (length payload))
+      (setf (cffi:mem-aref foreign-payload ':unsigned-char j)
+            (aref payload j)))
+    (pzmq:send socket foreign-identity :len (length identity) :sndmore t)
+    (when null-frame?
+      (pzmq:send socket "" :len 0 :sndmore t))
+    (pzmq:send socket foreign-payload :len (length payload) :sndmore nil))
+  nil)
+
+
+(define-condition not-an-RPCRequest (simple-error)
+  ((object :initarg :object :reader not-an-RPCRequest-object))
+  (:report (lambda (condition stream)
+             (format stream "Received bad object as inbound RPC request:~%~a~&"
+                     (not-an-RPCRequest-object condition)))))
+
+(define-condition unknown-RPC-method (simple-error)
+  ((method-name :initarg :method-name :reader unknown-RPC-method-name))
+  (:report (lambda (condition stream)
+             (format stream "Received request for method \"~a\", which is not known to us."
+                     (unknown-RPC-method-name condition)))))
+
+
+(defun %worker-routine (&key
```

Comment:
Humor me with a better name? (:

Reply:
i tried, but i'm not very good at this game